### DISCUSSION TITLE
Implementation of iRock

### DISCUSSION THREAD

**Discussion by jowgn** [Louisvdw/dbus-serialbattery#1108] [General]
I have just finished to write a driver for a iRock system. https://dilocker.com/wp-content/uploads/2022/04/Blockschaltbild-iRock212.pdf but as you see there is not necessarily current value (Shunt optional). At the moment I get calculation errors, probably because I don't provide a current value. Is there any idea what to do if there's no current value or is that just not supported?

**Comment by jowgn** [comment]
Maybe @mr-manuel or @Louisvdw could point me in a direction how to make that possible. Most Installs I know use a Smartshunt and would only be interested in having the rest of the BMS data in the VRM Portal.

**Comment by Louisvdw** [comment]
There is a minimum set of values that are required and current is one of them. Without it the system don't know the power usage and not limits can be applied. 
If you just want to show the cell and battery voltage data, then I suggest you set a dummy current of 0 so that it can display. In this case you should not this battery to control your GX system.

**Reply by jowgn** [reply]
Thank you so much for the fast answer. I'll try with 0A as the current value.

**Reply by jowgn** [reply]
This is sort of a special used case so would you recommend to commit this into the main project?

**Comment by Louisvdw** [comment]
If you think other people would/could use it as well then yes. If you used fixed assumtions it might need to be adjusted first so that it could work for any system. The optional shunt needs to be able to to talk to the driver as well. The ECS driver is similar in that the green meter and the LiPro cell devices are seperate, but for the driver to run you have to implement both.

**Comment by jowgn** [comment]
In this case, the implementation for the user would be the same because there is a Modbus value for the current already, but it would be zero then. So if no Shunt is connected, current value would always be zero if one is connected it would show the actual current. I'm only not too sure about how to communicate that the current value will be shown will only be shown if there is a Shunt connected. I would just assume that people probably not use it not as intended and actually use it with the current value in the system not only for monitoring reasons.

**Reply by Louisvdw** [reply]
That should be fine. If the shunt value goes through the BMS and is zero when no shunt is present but give a value when it is available that would be logical. A note can be added to the docs for this BMS to notify any users.

**Reply by jowgn** [reply]
OK, thank you so much for your support. I'll finish the first version of the driver. I'll implement one of the iRock hardware's and do a first pull request with that. Then I can add the other hardware types. Thank you again for your fast support.

**Comment by jowgn** [comment]
I guess the same goes for the Temp. Set it to something reasonable if there is no value provided by the BMS?

**Reply by mr-manuel** [reply]
If you check my repo, there is a well documented [`battery_template.py`](https://github.com/mr-manuel/venus-os_dbus-serialbattery/blob/master/etc/dbus-serialbattery/bms/battery_template.py).
